jbpf_tests/functional/report_stats/report_stats_hook_test.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * This test does the following:
+ * 1. It loads a single codelet (C1).
+ * 2. The codelet has a single hook (report_stats) that is called by the agent internally at the interval of
+ * MAINTENANCE_CHECK_INTERVAL (see jbpf_perf.c).
+ * 3. Then we check if the codelet (attached to report_stats hook) has been actually called by the agent. If it has been
+ * called, we should get the output from the codelet.
+ * 4. It uses the LCM IPC API to unload the codelet.
+ */
+
+#include <assert.h>
+#include <semaphore.h>
+#include <fcntl.h>
+#include <sys/wait.h>
+#include <signal.h>
+
+#include "jbpf.h"
+#include "jbpf_agent_common.h"
+#include "jbpf_utils.h"
+
+// Contains the struct and hook definitions
+#include "jbpf_test_def.h"
+
+pid_t cpid = -1;
+
+sem_t sem;
+
+jbpf_io_stream_id_t stream_id_c1 = {
+    .id = {0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xAA, 0xBB, 0xCC, 0xDD, 0xEE, 0xFF}};
+
+static void
+io_channel_check_output(jbpf_io_stream_id_t* stream_id, void** bufs, int num_bufs, void* ctx)
+{
+    int count = 0;
+    for (int i = 0; i < num_bufs; i++) {
+        if (memcmp(stream_id, &stream_id_c1, sizeof(stream_id_c1)) == 0) {
+            // Output from C1. Check that the counter has the expected value
+            count++;
+        } else {
+            // Unexpected output. Fail the test
+            assert(false);
+        }
+    }
+
+    // This means the report_stats hook has been called
+    if (count > 0) {
+        sem_post(&sem);
+    } else {
+        // we don't get any output from the codelet, meaning the report_stats hook has not been called
+        assert(false);
```

Comment:
Couldn't this test fail due to timing issues? What if the output thread is called right before the maintenance thread is called? Wouldn't `io_channel_check_output()` fail in this case although everything might be working fine?

Reply:
We use sem_post(lcm_ipc_sem) to make sure the codelet is loaded only after maintenance thread is created. By default, a dummy output_handler is registered, before `io_channel_check_output` is registered. 

I've ran the test 500 times, all passing.

Reply:
I don't think this is a problem. As the test use the semphore to ensure the maintenance thread is called see line 47 and 109.